src/Microsoft.Diagnostics.Runtime/src/Common/DACNameParser.cs

```diff
@@ -0,0 +1,831 @@
+                    {
+                        nextTargetFulfillmentIndex = i;
+                        break;
+                    }
+                }
+
+                Debug.Assert(nextTargetFulfillmentIndex != genericTargetIndex, "Ran out of args to back-propagate to satisfy generic requirements of earlier generic parameter.");
+                if (nextTargetFulfillmentIndex == genericTargetIndex)
+                    return false;
+
+                // Add the located argument as a generic arg to our previous generic type
+                targetTypeToPatch.AddGenericArg(genericArgs[nextTargetFulfillmentIndex]);
+
+                // Remove the back-propagaeted argument from our argument list since it is now contained within targetTypeToPatch
+                genericArgs.RemoveAt(nextTargetFulfillmentIndex);
+            }
+
+            // Patch our updated type info now that we have satisfied all of its generic arg requirements
+            genericArgs[genericTargetIndex] = targetTypeToPatch;
+
+            return true;
+        }
+
+        private static void UnifyNestedClasses(int parsingGenericArgListDepth, List<TypeNameSegment> genericArgs)
+        {
+            // Walk backwards so we can fold the multiple levels of nested classes backwards into the appropriate parent. It is important to do this BEFORE we propagate
+            // anything into the nameSegment list, since propagation from here to that list means patching a generic arg in that type, as such we want to patch the full
+            // (nested) type name, not the partial (outer parent) type name.
+            for (int i = genericArgs.Count - 1; i >= 1; i--)
+            {
+                TypeNameSegment candidateSegment = genericArgs[i];
+
+                // Once we hit an arg that came from a parsing arg depth < than the current, we can stop, we won't be processing any of those args yet, we will when we unwind
+                // that parsing level
+                if (candidateSegment.ParsingArgDepth < parsingGenericArgListDepth)
+                    break;
+
+                // Don't fold any that still have unfulfilled generics, leave them so they can be fulfilled
+                //
+                // TODO: Is this a thing? The need to check for HasUnfulfilledGenericArgs? The code before this that propagates generic args only fixes a single generic type target
+                // per pass...can there ever be more than one? The closing of a generic param list triggers this code, so the only way for there to be more than one generic type needing
+                // patching is EITHER a nested generic type in a generic parent or a generic param to a generic type, but the latter will trigger two seperate levels of generic arg parsing,
+                // the former is handled in the back-propagation code before this, and this collapses the nested generic (that has been resolved) into the parent generic (that may not
```

Comment:
Whoops, never took this out (the TODO) :P I think I landed on this was necessary (hence I left it in), at worst it isn't actively harmful. Can probably remove the TODO at the very least.